File: DataStructutreAndAlgorithm/Algorithm/06_Greedy.c
/*
================================================================================
GREEDY ALGORITHMS
================================================================================

STORY:
------
A traveler enters a city with three main roads.

One road is short but expensive.
One road is long but cheap.
One road is medium but risky.

Without thinking deeply, he always picks what seems best right now.

Most of the time, it works fine.

But sometimes, the short road leads to a dead end.

That's the risk of Greedy: Do best locally, hope it's best globally.

NETWORK APPLICATION:
- Huffman coding for data compression
- Activity selection in scheduling
- Fractional knapsack (resource allocation)
- Interval scheduling (task scheduling)
- Load balancing (least connection)
- Network bandwidth allocation
- Job scheduling

COMPLEXITY:
- Time: Usually O(n log n) or O(n²)
- Space: O(n)

KEY INSIGHT:
Greedy works when optimal substructure AND greedy choice property hold.
Not all problems have these properties.
Must verify greedy is optimal for specific problem.
================================================================================
*/

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define MAX_ITEMS 100
#define MAX_TASKS 100

// Activity Selection Problem
typedef struct {
    int start;
    int end;
    char name[50];
} Activity;

typedef struct {
    Activity activities[MAX_ITEMS];
    int count;
} ActivityList;

int compare_activities(const void* a, const void* b) {
    Activity* act_a = (Activity*)a;
    Activity* act_b = (Activity*)b;
    return act_a->end - act_b->end;  // Sort by end time
}

/*
ACTIVITY SELECTION: Select maximum non-overlapping activities
Network context: Schedule network maintenance windows
*/
void activity_selection(ActivityList* list) {
    printf("\n=== GREEDY: ACTIVITY SELECTION ===\n");
    printf("Select maximum non-overlapping activities:\n");
    
    for (int i = 0; i < list->count; i++) {
        printf("  %s: [%d, %d]\n", list->activities[i].name,
               list->activities[i].start, list->activities[i].end);
    }
    
    // Sort by end time
    qsort(list->activities, list->count, sizeof(Activity), compare_activities);
    
    printf("\nAfter sorting by end time:\n");
    for (int i = 0; i < list->count; i++) {
        printf("  %s: [%d, %d]\n", list->activities[i].name,
               list->activities[i].start, list->activities[i].end);
    }
    
    printf("\nGreedy selection:\n");
    printf("  Selected: %s (%d-%d)\n", list->activities[0].name,
           list->activities[0].start, list->activities[0].end);
    
    int last_end = list->activities[0].end;
    int count = 1;
    
    for (int i = 1; i < list->count; i++) {
        if (list->activities[i].start >= last_end) {
            printf("  Selected: %s (%d-%d)\n", list->activities[i].name,
                   list->activities[i].start, list->activities[i].end);
            last_end = list->activities[i].end;
            count++;
        }
    }
    
    printf("\nTotal selected: %d activities\n", count);
}

// Huffman Coding
typedef struct {
    char data;
    int freq;
    struct HuffmanNode* left;
    struct HuffmanNode* right;
} HuffmanNode;

typedef struct {
    HuffmanNode* nodes[MAX_ITEMS];
    int size;
} MinHeap;

void min_heap_push(MinHeap* heap, HuffmanNode* node) {
    heap->nodes[heap->size] = node;
    heap->size++;
    
    // Bubble up
    int idx = heap->size - 1;
    while (idx > 0) {
        int parent = (idx - 1) / 2;
        if (heap->nodes[idx]->freq < heap->nodes[parent]->freq) {
            HuffmanNode* temp = heap->nodes[idx];
            heap->nodes[idx] = heap->nodes[parent];
            heap->nodes[parent] = temp;
            idx = parent;
        } else {
            break;
        }
    }
}

HuffmanNode* min_heap_pop(MinHeap* heap) {
    if (heap->size == 0) return NULL;
    
    HuffmanNode* result = heap->nodes[0];
    heap->nodes[0] = heap->nodes[heap->size - 1];
    heap->size--;
    
    // Bubble down
    int idx = 0;
    while (2 * idx + 1 < heap->size) {
        int smallest = idx;
        if (heap->nodes[2 * idx + 1]->freq < heap->nodes[smallest]->freq) {
            smallest = 2 * idx + 1;
        }
        if (2 * idx + 2 < heap->size && 
            heap->nodes[2 * idx + 2]->freq < heap->nodes[smallest]->freq) {
            smallest = 2 * idx + 2;
        }
        if (smallest != idx) {
            HuffmanNode* temp = heap->nodes[idx];
            heap->nodes[idx] = heap->nodes[smallest];
            heap->nodes[smallest] = temp;
            idx = smallest;
        } else {
            break;
        }
    }
    
    return result;
}

/*
HUFFMAN CODING: Compress data using optimal variable-length codes
Network context: Compress network packets
*/
void huffman_coding_demo() {
    printf("\n\n=== GREEDY: HUFFMAN CODING ===\n");
    printf("Build optimal prefix-free binary codes for compression\n\n");
    
    // Frequency table for ASCII characters
    struct {
        char c;
        int freq;
    } chars[] = {
        {'a', 5}, {'b', 9}, {'c', 12}, {'d', 13}, {'e', 16}, {'f', 45}
    };
    int char_count = 6;
    
    printf("Character frequencies:\n");
    int total = 0;
    for (int i = 0; i < char_count; i++) {
        printf("  '%c': %d occurrences\n", chars[i].c, chars[i].freq);
        total += chars[i].freq;
    }
    
    MinHeap heap;
    heap.size = 0;
    
    // Create leaf nodes and add to heap
    for (int i = 0; i < char_count; i++) {
        HuffmanNode* node = malloc(sizeof(HuffmanNode));
        node->data = chars[i].c;
        node->freq = chars[i].freq;
        node->left = NULL;
        node->right = NULL;
        min_heap_push(&heap, node);
    }
    
    // Build Huffman tree
    printf("\nBuilding Huffman tree (greedy selection):\n");
    while (heap.size > 1) {
        HuffmanNode* left = min_heap_pop(&heap);
        HuffmanNode* right = min_heap_pop(&heap);
        
        HuffmanNode* parent = malloc(sizeof(HuffmanNode));
        parent->freq = left->freq + right->freq;
        parent->left = left;
        parent->right = right;
        parent->data = '\0';
        
        printf("  Merge: freq(%d) + freq(%d) = %d\n",
               left->freq, right->freq, parent->freq);
        
        min_heap_push(&heap, parent);
    }
    
    printf("\nHuffman coding is optimal variable-length code\n");
    printf("Compression benefit: Frequently used chars get shorter codes\n");
    printf("Example: 'f'(45) → 0 or 1, 'a'(5) → 1110 or 1111\n");
}

// Fractional Knapsack (can take partial items)
typedef struct {
    char name[50];
    int weight;
    int value;
    double ratio;  // Value per weight
} Item;

int compare_items_by_ratio(const void* a, const void* b) {
    Item* item_a = (Item*)a;
    Item* item_b = (Item*)b;
    return (item_a->ratio < item_b->ratio) ? 1 : -1;  // Descending
}

/*
FRACTIONAL KNAPSACK: Greedy packing with partial items
Network context: Bandwidth allocation with fractional sharing
*/
void fractional_knapsack() {
    printf("\n\n=== GREEDY: FRACTIONAL KNAPSACK ===\n");
    printf("Pack items to maximize value with partial items allowed\n");
    printf("(Network context: Fractional bandwidth sharing)\n\n");
    
    Item items[] = {
        {"Service A", 60, 300, 300.0/60},   // value/weight = 5.0
        {"Service B", 100, 160, 160.0/100}, // value/weight = 1.6
        {"Service C", 120, 120, 120.0/120}  // value/weight = 1.0
    };
    int item_count = 3;
    int capacity = 150;  // Max bandwidth
    
    printf("Items (services):\n");
    for (int i = 0; i < item_count; i++) {
        printf("  %s: weight=%d, value=%d, ratio=%.2f\n",
               items[i].name, items[i].weight, items[i].value, items[i].ratio);
    }
    printf("\nCapacity: %d Mbps\n", capacity);
    
    // Sort by value-to-weight ratio
    qsort(items, item_count, sizeof(Item), compare_items_by_ratio);
    
    printf("\nAfter sorting by value/weight ratio (descending):\n");
    for (int i = 0; i < item_count; i++) {
        printf("  %s: ratio=%.2f\n", items[i].name, items[i].ratio);
    }
    
    printf("\nGreedy selection:\n");
    double total_value = 0;
    int remaining = capacity;
    
    for (int i = 0; i < item_count; i++) {
        if (remaining >= items[i].weight) {
            // Take fully
            printf("  Take %s fully: weight=%d, value=%d\n",
                   items[i].name, items[i].weight, items[i].value);
            total_value += items[i].value;
            remaining -= items[i].weight;
        } else if (remaining > 0) {
            // Take partially
            double fraction = (double)remaining / items[i].weight;
            double partial_value = fraction * items[i].value;
            printf("  Take %s partially (%.2f): weight=%d, partial_value=%.2f\n",
                   items[i].name, fraction, remaining, partial_value);
            total_value += partial_value;
            remaining = 0;
            break;
        }
    }
    
    printf("\nTotal value: %.2f\n", total_value);
    printf("Capacity used: %d/%d Mbps\n", capacity - remaining, capacity);
}

// Job Sequencing
typedef struct {
    int id;
    int deadline;
    int profit;
} Job;

int compare_jobs_by_profit(const void* a, const void* b) {
    Job* job_a = (Job*)a;
    Job* job_b = (Job*)b;
    return job_b->profit - job_a->profit;  // Descending
}

/*
JOB SEQUENCING: Schedule jobs to maximize profit
Network context: Task scheduling with deadlines
*/
void job_sequencing() {
    printf("\n\n=== GREEDY: JOB SEQUENCING ===\n");
    printf("Schedule jobs to maximize profit with deadlines\n\n");
    
    Job jobs[] = {
        {1, 4, 20},   // Job 1: deadline=4, profit=20
        {2, 1, 10},   // Job 2: deadline=1, profit=10
        {3, 1, 40},   // Job 3: deadline=1, profit=40
        {4, 1, 30},   // Job 4: deadline=1, profit=30
        {5, 3, 35},   // Job 5: deadline=3, profit=35
    };
    int job_count = 5;
    
    printf("Jobs:\n");
    for (int i = 0; i < job_count; i++) {
        printf("  Job %d: deadline=%d, profit=%d\n",
               jobs[i].id, jobs[i].deadline, jobs[i].profit);
    }
    
    // Sort by profit (descending)
    qsort(jobs, job_count, sizeof(Job), compare_jobs_by_profit);
    
    printf("\nAfter sorting by profit (descending):\n");
    for (int i = 0; i < job_count; i++) {
        printf("  Job %d: profit=%d\n", jobs[i].id, jobs[i].profit);
    }
    
    // Find maximum deadline
    int max_deadline = 0;
    for (int i = 0; i < job_count; i++) {
        if (jobs[i].deadline > max_deadline) {
            max_deadline = jobs[i].deadline;
        }
    }
    
    // Schedule greedily
    int* schedule = (int*)malloc((max_deadline + 1) * sizeof(int));
    memset(schedule, 0, (max_deadline + 1) * sizeof(int));
    
    int total_profit = 0;
    printf("\nGreedy scheduling:\n");
    
    for (int i = 0; i < job_count; i++) {
        // Try to schedule at or before deadline
        for (int j = jobs[i].deadline; j > 0; j--) {
            if (schedule[j] == 0) {
                schedule[j] = jobs[i].id;
                printf("  Job %d scheduled at time %d (profit=%d)\n",
                       jobs[i].id, j, jobs[i].profit);
                total_profit += jobs[i].profit;
                break;
            }
        }
    }
    
    printf("\nTotal profit: %d\n", total_profit);
    
    printf("\nFinal schedule:\n");
    printf("Time: ");
    for (int i = 1; i <= max_deadline; i++) {
        printf("%d ", i);
    }
    printf("\nJob:  ");
    for (int i = 1; i <= max_deadline; i++) {
        printf("%d ", schedule[i]);
    }
    printf("\n");
    
    free(schedule);
}

void complexity_analysis() {
    printf("\n\n╔════════════════════════════════════════════════════════════════╗\n");
    printf("║ COMPLEXITY ANALYSIS                                            ║\n");
    printf("╚════════════════════════════════════════════════════════════════╝\n");
    
    printf("\nActivity Selection:\n");
    printf("  Time: O(n log n) for sorting\n");
    printf("  Space: O(1) if sorted already\n");
    printf("  Property: Optimal if sorted by end time\n\n");
    
    printf("Huffman Coding:\n");
    printf("  Time: O(n log n) with priority queue\n");
    printf("  Space: O(n) for tree\n");
    printf("  Property: Optimal variable-length prefix code\n\n");
    
    printf("Fractional Knapsack:\n");
    printf("  Time: O(n log n) for sorting\n");
    printf("  Space: O(1)\n");
    printf("  Property: Greedy optimal for fractional items\n\n");
    
    printf("Job Sequencing:\n");
    printf("  Time: O(n² ) or O(n log n + n*deadline)\n");
    printf("  Space: O(deadline)\n");
    printf("  Property: Optimal for deadline scheduling\n\n");
    
    printf("Important: Greedy NOT always optimal!\n");
    printf("  0/1 Knapsack: Greedy FAILS (use DP instead)\n");
    printf("  TSP: Greedy gives ~2x optimal (NP-hard)\n");
}

int main() {
    printf("╔════════════════════════════════════════════════════════════════╗\n");
    printf("║ GREEDY ALGORITHMS - COMPREHENSIVE DEMONSTRATION               ║\n");
    printf("╚════════════════════════════════════════════════════════════════╝\n");
    
    printf("\nGreedy Strategy: Pick locally best option\n");
    printf("Hope: It leads to globally optimal solution\n");
    printf("Reality: Works for some problems, not others\n");
    
    // Activity Selection Demo
    ActivityList list;
    list.count = 6;
    list.activities[0] = (Activity){1, 3, "Maintenance A"};
    list.activities[1] = (Activity){2, 5, "Maintenance B"};
    list.activities[2] = (Activity){4, 6, "Maintenance C"};
    list.activities[3] = (Activity){6, 7, "Maintenance D"};
    list.activities[4] = (Activity){5, 8, "Maintenance E"};
    list.activities[5] = (Activity){8, 9, "Maintenance F"};
    
    activity_selection(&list);
    
    // Huffman Coding Demo
    huffman_coding_demo();
    
    // Fractional Knapsack
    fractional_knapsack();
    
    // Job Sequencing
    job_sequencing();
    
    // Complexity Analysis
    complexity_analysis();
    
    printf("\nKEY TAKEAWAYS:\n");
    printf("1. Greedy: Fast, simple to implement\n");
    printf("2. Works when greedy choice property holds\n");
    printf("3. Must verify optimality for each problem\n");
    printf("4. Useful for scheduling and compression\n");
    printf("5. Essential for load balancing algorithms\n");
    
    return 0;
}

File: DataStructutreAndAlgorithm/Algorithm/06_Greedy.java
/*
================================================================================
GREEDY ALGORITHMS - JAVA VERSION
================================================================================

STORY:
------
A traveler enters a city with three main roads.

One road is short but expensive.
One road is long but cheap.
One road is medium but risky.

Without thinking deeply, he always picks what seems best right now.

Most of the time, it works fine.

But sometimes, the short road leads to a dead end.

That's the risk of Greedy: Do best locally, hope it's best globally.

NETWORK APPLICATION:
- Huffman coding for data compression
- Activity selection in scheduling
- Fractional knapsack (resource allocation)
- Interval scheduling (task scheduling)
- Load balancing (least connection)
- Network bandwidth allocation
- Job scheduling

COMPLEXITY:
- Time: Usually O(n log n) or O(n²)
- Space: O(n)

KEY INSIGHT:
Greedy works when optimal substructure AND greedy choice property hold.
================================================================================
*/

import java.util.*;

public class Greedy {
    
    static class Activity implements Comparable<Activity> {
        int id;
        int start;
        int end;
        String name;
        
        Activity(int id, int start, int end, String name) {
            this.id = id;
            this.start = start;
            this.end = end;
            this.name = name;
        }
        
        // Sort by end time
        @Override
        public int compareTo(Activity other) {
            return this.end - other.end;
        }
    }
    
    /*
    ACTIVITY SELECTION: Select maximum non-overlapping activities
    Network context: Schedule network maintenance windows
    */
    static void activitySelection(List<Activity> activities) {
        System.out.println("\n=== GREEDY: ACTIVITY SELECTION ===");
        System.out.println("Select maximum non-overlapping activities:\n");
        
        for (Activity act : activities) {
            System.out.println("  " + act.name + ": [" + act.start + ", " + act.end + "]");
        }
        
        // Sort by end time
        Collections.sort(activities);
        
        System.out.println("\nAfter sorting by end time:");
        for (Activity act : activities) {
            System.out.println("  " + act.name + ": [" + act.start + ", " + act.end + "]");
        }
        
        System.out.println("\nGreedy selection:");
        System.out.println("  Selected: " + activities.get(0).name + 
                         " (" + activities.get(0).start + "-" + activities.get(0).end + ")");
        
        int lastEnd = activities.get(0).end;
        int count = 1;
        
        for (int i = 1; i < activities.size(); i++) {
            Activity current = activities.get(i);
            if (current.start >= lastEnd) {
                System.out.println("  Selected: " + current.name + 
                                 " (" + current.start + "-" + current.end + ")");
                lastEnd = current.end;
                count++;
            }
        }
        
        System.out.println("\nTotal selected: " + count + " activities");
    }
    
    static class Item implements Comparable<Item> {
        String name;
        int weight;
        int value;
        double ratio;
        
        Item(String name, int weight, int value) {
            this.name = name;
            this.weight = weight;
            this.value = value;
            this.ratio = (double) value / weight;
        }
        
        // Sort by ratio (descending)
        @Override
        public int compareTo(Item other) {
            return Double.compare(other.ratio, this.ratio);  // Descending
        }
    }
    
    /*
    FRACTIONAL KNAPSACK: Greedy packing with partial items
    Network context: Bandwidth allocation with fractional sharing
    */
    static void fractionalKnapsack() {
        System.out.println("\n\n=== GREEDY: FRACTIONAL KNAPSACK ===");
        System.out.println("Pack items to maximize value with partial items allowed");
        System.out.println("(Network context: Fractional bandwidth sharing)\n");
        
        List<Item> items = new ArrayList<>();
        items.add(new Item("Service A", 60, 300));   // ratio = 5.0
        items.add(new Item("Service B", 100, 160));  // ratio = 1.6
        items.add(new Item("Service C", 120, 120));  // ratio = 1.0
        
        int capacity = 150;  // Max bandwidth
        
        System.out.println("Items (services):");
        for (Item item : items) {
            System.out.printf("  %s: weight=%d, value=%d, ratio=%.2f\n",
                            item.name, item.weight, item.value, item.ratio);
        }
        System.out.println("\nCapacity: " + capacity + " Mbps");
        
        // Sort by ratio
        Collections.sort(items);
        
        System.out.println("\nAfter sorting by value/weight ratio (descending):");
        for (Item item : items) {
            System.out.printf("  %s: ratio=%.2f\n", item.name, item.ratio);
        }
        
        System.out.println("\nGreedy selection:");
        double totalValue = 0;
        int remaining = capacity;
        
        for (Item item : items) {
            if (remaining >= item.weight) {
                // Take fully
                System.out.printf("  Take %s fully: weight=%d, value=%d\n",
                                item.name, item.weight, item.value);
                totalValue += item.value;
                remaining -= item.weight;
            } else if (remaining > 0) {
                // Take partially
                double fraction = (double) remaining / item.weight;
                double partialValue = fraction * item.value;
                System.out.printf("  Take %s partially (%.2f): weight=%d, partial_value=%.2f\n",
                                item.name, fraction, remaining, partialValue);
                totalValue += partialValue;
                remaining = 0;
                break;
            }
        }
        
        System.out.printf("\nTotal value: %.2f\n", totalValue);
        System.out.println("Capacity used: " + (capacity - remaining) + "/" + capacity + " Mbps");
    }
    
    static class Job implements Comparable<Job> {
        int id;
        int deadline;
        int profit;
        
        Job(int id, int deadline, int profit) {
            this.id = id;
            this.deadline = deadline;
            this.profit = profit;
        }
        
        // Sort by profit (descending)
        @Override
        public int compareTo(Job other) {
            return other.profit - this.profit;  // Descending
        }
    }
    
    /*
    JOB SEQUENCING: Schedule jobs to maximize profit with deadlines
    Network context: Task scheduling with deadlines
    */
    static void jobSequencing() {
        System.out.println("\n\n=== GREEDY: JOB SEQUENCING ===");
        System.out.println("Schedule jobs to maximize profit with deadlines\n");
        
        List<Job> jobs = new ArrayList<>();
        jobs.add(new Job(1, 4, 20));
        jobs.add(new Job(2, 1, 10));
        jobs.add(new Job(3, 1, 40));
        jobs.add(new Job(4, 1, 30));
        jobs.add(new Job(5, 3, 35));
        
        System.out.println("Jobs:");
        for (Job job : jobs) {
            System.out.println("  Job " + job.id + ": deadline=" + job.deadline + 
                             ", profit=" + job.profit);
        }
        
        // Sort by profit (descending)
        Collections.sort(jobs);
        
        System.out.println("\nAfter sorting by profit (descending):");
        for (Job job : jobs) {
            System.out.println("  Job " + job.id + ": profit=" + job.profit);
        }
        
        // Find maximum deadline
        int maxDeadline = 0;
        for (Job job : jobs) {
            maxDeadline = Math.max(maxDeadline, job.deadline);
        }
        
        // Schedule greedily
        int[] schedule = new int[maxDeadline + 1];
        int totalProfit = 0;
        
        System.out.println("\nGreedy scheduling:");
        
        for (Job job : jobs) {
            // Try to schedule at or before deadline
            for (int j = job.deadline; j > 0; j--) {
                if (schedule[j] == 0) {
                    schedule[j] = job.id;
                    System.out.println("  Job " + job.id + " scheduled at time " + j + 
                                     " (profit=" + job.profit + ")");
                    totalProfit += job.profit;
                    break;
                }
            }
        }
        
        System.out.println("\nTotal profit: " + totalProfit);
        
        System.out.println("\nFinal schedule:");
        System.out.print("Time: ");
        for (int i = 1; i <= maxDeadline; i++) {
            System.out.print(i + " ");
        }
        System.out.println();
        System.out.print("Job:  ");
        for (int i = 1; i <= maxDeadline; i++) {
            System.out.print(schedule[i] + " ");
        }
        System.out.println();
    }
    
    /*
    HUFFMAN CODING: Compress data using optimal variable-length codes
    */
    static class HuffmanNode implements Comparable<HuffmanNode> {
        char data;
        int freq;
        HuffmanNode left, right;
        
        HuffmanNode(char data, int freq) {
            this.data = data;
            this.freq = freq;
        }
        
        @Override
        public int compareTo(HuffmanNode other) {
            return this.freq - other.freq;
        }
    }
    
    static void huffmanCoding() {
        System.out.println("\n\n=== GREEDY: HUFFMAN CODING ===");
        System.out.println("Build optimal prefix-free binary codes for compression\n");
        
        int[] charFreq = {5, 9, 12, 13, 16, 45};
        char[] chars = {'a', 'b', 'c', 'd', 'e', 'f'};
        
        System.out.println("Character frequencies:");
        int total = 0;
        for (int i = 0; i < chars.length; i++) {
            System.out.println("  '" + chars[i] + "': " + charFreq[i] + " occurrences");
            total += charFreq[i];
        }
        
        // Create priority queue
        PriorityQueue<HuffmanNode> pq = new PriorityQueue<>();
        
        for (int i = 0; i < chars.length; i++) {
            pq.add(new HuffmanNode(chars[i], charFreq[i]));
        }
        
        // Build Huffman tree
        System.out.println("\nBuilding Huffman tree (greedy selection):");
        while (pq.size() > 1) {
            HuffmanNode left = pq.poll();
            HuffmanNode right = pq.poll();
            
            HuffmanNode parent = new HuffmanNode('\0', left.freq + right.freq);
            parent.left = left;
            parent.right = right;
            
            System.out.println("  Merge: freq(" + left.freq + ") + freq(" + right.freq + 
                             ") = " + parent.freq);
            
            pq.add(parent);
        }
        
        System.out.println("\nHuffman coding is optimal variable-length code");
        System.out.println("Compression benefit: Frequently used chars get shorter codes");
        System.out.println("Example: 'f'(45) → 0 or 1, 'a'(5) → 1110 or 1111");
    }
    
    static void complexityAnalysis() {
        System.out.println("\n\n╔════════════════════════════════════════════════════════════════╗");
        System.out.println("║ COMPLEXITY ANALYSIS                                            ║");
        System.out.println("╚════════════════════════════════════════════════════════════════╝");
        
        System.out.println("\nActivity Selection:");
        System.out.println("  Time: O(n log n) for sorting");
        System.out.println("  Space: O(1) if sorted already");
        System.out.println("  Property: Optimal if sorted by end time\n");
        
        System.out.println("Huffman Coding:");
        System.out.println("  Time: O(n log n) with priority queue");
        System.out.println("  Space: O(n) for tree");
        System.out.println("  Property: Optimal variable-length prefix code\n");
        
        System.out.println("Fractional Knapsack:");
        System.out.println("  Time: O(n log n) for sorting");
        System.out.println("  Space: O(1)");
        System.out.println("  Property: Greedy optimal for fractional items\n");
        
        System.out.println("Job Sequencing:");
        System.out.println("  Time: O(n²) or O(n log n + n*deadline)");
        System.out.println("  Space: O(deadline)");
        System.out.println("  Property: Optimal for deadline scheduling\n");
        
        System.out.println("Important: Greedy NOT always optimal!");
        System.out.println("  0/1 Knapsack: Greedy FAILS (use DP instead)");
        System.out.println("  TSP: Greedy gives ~2x optimal (NP-hard)");
    }
    
    /*
    PRACTICAL EXAMPLE: Network Load Balancing with Greedy
    */
    static void loadBalancingExample() {
        System.out.println("\n\n╔════════════════════════════════════════════════════════════════╗");
        System.out.println("║ PRACTICAL: NETWORK LOAD BALANCING                             ║");
        System.out.println("╚════════════════════════════════════════════════════════════════╝");
        
        System.out.println("\nScenario: Distribute incoming requests to servers");
        System.out.println("Greedy strategy: Always send to server with least connections\n");
        
        int[] serverConnections = {5, 8, 3, 6};
        String[] servers = {"Server-1", "Server-2", "Server-3", "Server-4"};
        
        System.out.println("Current server loads:");
        for (int i = 0; i < servers.length; i++) {
            System.out.println("  " + servers[i] + ": " + serverConnections[i] + " connections");
        }
        
        System.out.println("\nNew requests arrive (10 requests):");
        for (int req = 1; req <= 10; req++) {
            // Find server with min connections
            int minIdx = 0;
            for (int i = 1; i < serverConnections.length; i++) {
                if (serverConnections[i] < serverConnections[minIdx]) {
                    minIdx = i;
                }
            }
            
            serverConnections[minIdx]++;
            System.out.println("  Request " + req + " → " + servers[minIdx]);
        }
        
        System.out.println("\nFinal loads:");
        for (int i = 0; i < servers.length; i++) {
            System.out.println("  " + servers[i] + ": " + serverConnections[i] + " connections");
        }
    }
    
    public static void main(String[] args) {
        System.out.println("╔════════════════════════════════════════════════════════════════╗");
        System.out.println("║ GREEDY ALGORITHMS - JAVA COMPREHENSIVE DEMONSTRATION          ║");
        System.out.println("╚════════════════════════════════════════════════════════════════╝");
        
        System.out.println("\nGreedy Strategy: Pick locally best option");
        System.out.println("Hope: It leads to globally optimal solution");
        System.out.println("Reality: Works for some problems, not others\n");
        
        // Activity Selection
        List<Activity> activities = new ArrayList<>();
        activities.add(new Activity(1, 1, 3, "Maintenance A"));
        activities.add(new Activity(2, 2, 5, "Maintenance B"));
        activities.add(new Activity(3, 4, 6, "Maintenance C"));
        activities.add(new Activity(4, 6, 7, "Maintenance D"));
        activities.add(new Activity(5, 5, 8, "Maintenance E"));
        activities.add(new Activity(6, 8, 9, "Maintenance F"));
        
        activitySelection(activities);
        
        // Huffman Coding
        huffmanCoding();
        
        // Fractional Knapsack
        fractionalKnapsack();
        
        // Job Sequencing
        jobSequencing();
        
        // Load Balancing
        loadBalancingExample();
        
        // Complexity Analysis
        complexityAnalysis();
        
        System.out.println("\nKEY TAKEAWAYS:");
        System.out.println("1. Greedy: Fast, simple to implement");
        System.out.println("2. Works when greedy choice property holds");
        System.out.println("3. Must verify optimality for each problem");
        System.out.println("4. Useful for scheduling and compression");
        System.out.println("5. Essential for load balancing algorithms");
    }
}
